### PR TITLE
Start VM entrypoint when flag is set in Start command

### PR DESCRIPTION
**Pull request by evik42**
Starts the entrypoint defined in the metadata according to docker rules:
- if ENTRYPOINT or CMD is present but the other is missing, the command defined by the setting will be executed
- if both is present then CMD is concatenated to ENTRYPOINT to use it as parameter(s) and the extended command is executed

If the `start_entrypoint` flag is defined for Start command but the metadata does not define an entrypoint then an error will be returned without starting the VM.

Solves #93 
Resolves https://github.com/golemfactory/yagna/issues/1350

### CODE REVIEW

File: runtime/src/main.rs [hunk 1]
```diff
@@ -437,6 +461,49 @@ fn convert_result<T>(
...
 }
 
+fn extract_entrypoint(config: &ContainerConfig) -> Option<Vec<String>> {
+    let mut entrypoint = config.entrypoint.clone().unwrap_or_default();
+    entrypoint.append(&mut config.cmd.clone().unwrap_or_default());
+    if entrypoint.is_empty() {
+        None
+    } else {
+        Some(entrypoint)
+    }
```
Comment (mfranciszkiewicz) [inline runtime/src/main.rs:471]: cosmetic change suggestion

```suggestion
    config.entrypoint.clone().map(|mut entrypoint| {
        entrypoint.extend(config.cmd.clone().unwrap_or_default());
        entrypoint
    })
```

File: runtime/src/main.rs [hunk 2]
```diff
@@ -437,6 +461,49 @@ fn convert_result<T>(
...
+            args,
+            ..Default::default()
+        },
+    )
+    .await
+    .map(|pid| {
+        use serde_json::json;
+
+        json!({
+            "start": if start_response.is_some() { start_response.unwrap() } else { json!(null) },
```
Comment (mfranciszkiewicz) [inline runtime/src/main.rs:501]:
```suggestion
            "start": start_response.unwrap_or(json!(null)),
```

File: runtime/src/main.rs [hunk 3]
```diff
@@ -117,9 +132,18 @@ impl ya_runtime_sdk::Runtime for Runtime {
...
+            let start_result = start(workdir, data.clone(), emitter).await;
+
+            if start_result.is_ok() && entrypoint.is_some() {
+                run_entrypoint(start_result.unwrap(), entrypoint.unwrap(), data)
+                    .await
+                    .map(Some)
+                    .map_err(Into::into)
+            } else {
+                start_result.map_err(Into::into)
+            }
```
Comment (mfranciszkiewicz) [inline runtime/src/main.rs:145]:
```suggestion
            let start_result = start(workdir, data.clone(), emitter).await?;
            Ok(match entrypoint {
                Some(entrypoint) => Some(run_entrypoint(start_result, entrypoint, data).await?),
                None => start_result,
            })
```

File: runtime/src/main.rs (hunk 1 above)
Comment (evik42) [reply runtime/src/main.rs:471]: Your suggestion would only work if entrypoint is defined. Docker would run CMD as entrypoint if entrypoint is not defined.

Comment (evik42) [reply runtime/src/main.rs:471]: Changing `append` to `extend`.

Comment (mfranciszkiewicz) [reply runtime/src/main.rs:471]: You’re right, we we would still need to .or() cmd

Comment (evik42) [reply runtime/src/main.rs:471]: @mfranciszkiewicz does this look better?
    let entrypoint = config.entrypoint.clone().unwrap_or_default().into_iter()
        .chain(config.cmd.clone().unwrap_or_default())
        .collect::<Vec<_>>();

**Review by mfranciszkiewicz — APPROVED** [review]

**Pull request merged**